README.md
# DateConvert

DateConvert is a package for the Dart language used to perform date conversions

Abstracts the logic of converting between dates of different types (or the same type, such as converting from String 'dd/MM/yy' to 'yyyy-MM-dd,' for example), requiring only the input date and the expected format after conversion

![](https://github.com/Massau/flutter_date_convert/blob/main/assets/date_convert.jpg)

### Install

This will add a line like this to your package's pubspec.yaml (and run an implicit dart pub get)

```yaml
  dependencies:
    date_convert:
      git:
        url: https://github.com/Massau/flutter_date_convert.git
        ref: '0.0.1' # branch name
```

### Getting Started

Provide a date in the form of DateTime, int, or String as input and obtain the converted date in the desired data type.

Throw an exception if an invalid value is provided for any parameter.

If the input and output date types are the same, no conversion will occur, except for String and String, when the input date format differs from the expected format.


### Parameters

- date is required and can be provided in the following formats: DateTime, int, or String
- expectedTypeOrFormat is required. If you want the converted date to be a String, you can either provide the String class as the parameter or specify the desired format for the String after conversion. If the class is provided without specifying the format, a default DateFormat will be assigned
- currentStringFormat is required when the input date is of type String. Dates in other formats will not be affected. The following options are valid to inform in this parameter: 'dmy', 'mdy', 'ymd'. Providing the incorrect order of day, month, and year in the String may result in an incorrect output or an exception
- locale is optional and can be provided or omitted when the output type is String. Different output types will not be affected by this parameter. If nothing is provided, the default value is assumed to be 'en_US'


### Conversions

This package can perform the following date conversions:

#### String to String Conversion (with different formats)

The input date has the format "yMd", and the expected output date format is 'yMMMMd'.

```dart
  Parser _parser = Parser();
  
  Widget build(BuildContext context) {
    return Text(
      _parser.dateConvert(
        '25/06/2023',
        expectedTypeOrFormat: 'yMMMMd',
        currentStringFormat: 'dmy',
      ),
      textAlign: TextAlign.center,
    );
  }
```

The code snippet above produces the following result:

```
'June 25, 2023'
```

#### String to String Conversion (Specifying Locale as a parameter)

It is possible to specify the locale of the date after conversion.

This parameter is optional and can be provided when the output type is String. Different output types will not be affected by it.

If the parameter is not provided for a String conversion, the default value is 'en_US'. Below is an example that specifies the locale:

```dart
  Parser _parser = Parser();
  
  Widget build(BuildContext context) {
    return Text(
      _parser.dateConvert(
        '25/06/2023',
        expectedTypeOrFormat: 'yMMMMd',
        currentStringFormat: 'dmy',
        locale: 'pt_BR'
      ),
      textAlign: TextAlign.center,
    );
  }
```

The code snippet above produces the following result:


```
'25 de junho de 2023'
```

To identify if a locale is valid, consult [this link](https://pub.dev/documentation/intl/latest/date_symbol_data_http_request/availableLocalesForDateFormatting.html)

#### int to int Conversion

No conversion is performed when the data types are the same, and the number itself is returned:

```dart
  Parser _parser = Parser();
  
  Widget build(BuildContext context) {
    return Text(
      _parser.dateConvert(
        1687662000000,
        expectedTypeOrFormat: int
      ).toString(),
      textAlign: TextAlign.center,
    );
  }
```

The code snippet above produces the following result:

```
1687662000000
```

#### DateTime to int Conversion

```dart
  Parser _parser = Parser();
  
  Widget build(BuildContext context) {
    return Text(
      _parser.dateConvert(
        DateTime(2023, 06, 24),
        expectedTypeOrFormat: int
      ).toString(),
      textAlign: TextAlign.center,
    );
  }
```

The code snippet above produces the following result:

```
1687575600000"
```


CHANGELOG.md
## 0.0.1

Perform date conversions


test/flutter_date_convert_test.dart
import 'package:flutter_date_convert/flutter_date_convert.dart';
import 'package:flutter_date_convert/src/parser.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Format Class', () {
    test('Valid dates', () {
      final Format format = Format();

      expect(format.validDateTime(DateTime(2023, 9, 30)), true);
      expect(format.validDateTime(DateTime(2023, 12, 31)), true);
    });
  });

  group('Parser', () {
    test('Convert dates', () {
      Parser _parser = Parser();

      expect(
        _parser.dateConvert(DateTime(2023, 06, 24),
            expectedTypeOrFormat: DateTime),
        DateTime(2023, 06, 24),
      );
      expect(
        _parser.dateConvert(1687575600000, expectedTypeOrFormat: DateTime),
        DateTime(2023, 06, 24),
      );
      expect(
        _parser.dateConvert(
          1687575600000,
          expectedTypeOrFormat: DateTime,
          currentStringFormat: 'dmy',
        ),
        DateTime(2023, 06, 24),
      );
      expect(
        _parser.dateConvert(
          1687575600000,
          expectedTypeOrFormat: DateTime,
          currentStringFormat: 'dmy',
          locale: 'pt_BR',
        ),
        DateTime(2023, 06, 24),
      );
      expect(
        _parser.dateConvert(1687662000000,
            expectedTypeOrFormat: 'yMd', currentStringFormat: 'dmy'),
        '6/25/2023',
      );
      expect(
        _parser.dateConvert(1687662000000, expectedTypeOrFormat: int),
        1687662000000,
      );
      expect(
        _parser.dateConvert(
          1687662000000,
          expectedTypeOrFormat: 'yMd',
          currentStringFormat: 'dmy',
          locale: 'pt_BR',
        ),
        '25/06/2023',
      );
      expect(
        _parser.dateConvert('25/06/2023',
            expectedTypeOrFormat: 'yMd', currentStringFormat: 'dmy'),
        '6/25/2023',
      );
      expect(
        _parser.dateConvert(
          '25/06/2023',
          expectedTypeOrFormat: 'yMd',
          currentStringFormat: 'dmy',
          locale: 'pt_BR',
        ),
        '25/06/2023',
      );
      expect(
        _parser.dateConvert('25/06/2023',
            expectedTypeOrFormat: 'yMMMMd',
            currentStringFormat: 'dmy',
            locale: 'pt_BR'),
        '25 de junho de 2023',
      );
      expect(
        _parser.dateConvert('25/06/2023',
            expectedTypeOrFormat: String, currentStringFormat: 'dmy'),
        '6/25/2023',
      );
      expect(
        _parser.dateConvert(DateTime(2023, 06, 25),
            expectedTypeOrFormat: String),
        '6/25/2023',
      );
      expect(
        _parser.dateConvert(1687662000000, expectedTypeOrFormat: String),
        '6/25/2023',
      );
    });
  });
}


lib/src/format.dart
import 'package:flutter_date_convert/src/convert.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:intl/intl.dart';

import 'utils.dart';

class Format {
  typeOrFormatIsString(typeOrFormat) =>
      typeOrFormat == String || typeOrFormat is String;

  _regularExpression(date) {
    bool _regularExpressionValid = false;
    var _amountCharacters = {
      'day': '[0-3][0-9]',
      'month': '[01][1-9]',
      'year': '[0-9]{4}',
    };
    List<RegExp> _dateFormats = [
      RegExp(
          '^${_amountCharacters["day"]}[-\\\/]${_amountCharacters["month"]}[-\\\/]${_amountCharacters["year"]}\$'),
      RegExp(
          '^${_amountCharacters["year"]}[-\\\/]${_amountCharacters["month"]}[-\\\/]${_amountCharacters["day"]}\$'),
    ];

    for (var i = 0; i < _dateFormats.length; i++) {
      if (_dateFormats[i].hasMatch(date)) {
        _regularExpressionValid = true;
        break;
      }
    }

    return _regularExpressionValid;
  }

  _validStringDay(List date, String format) {
    Convert _convert = Convert();
    format = format.toString().toLowerCase();
    bool _valid = false;

    if (format == 'dmy') {
      _valid = _convert.fromInt(date[0]) > 0 &&
          _convert.fromInt(date[0]) <=
              mounthsNonLeapYear[_convert.fromInt(date[1])]['days'];
    }

    if (format == 'ymd') {
      _valid = _convert.fromInt(date[2]) > 0 &&
          _convert.fromInt(date[2]) <=
              mounthsNonLeapYear[_convert.fromInt(date[1])]['days'];
    }

    if (format == 'mdy') {
      _valid = _convert.fromInt(date[1]) > 0 &&
          _convert.fromInt(date[1]) <=
              mounthsNonLeapYear[_convert.fromInt(date[1])]['days'];
    }

    return _valid;
  }

  _validStringMonth(month) {
    Convert _convert = Convert();

    return _convert.fromInt(month) > 0 && _convert.fromInt(month) <= 12;
  }

  bool _validString(date, currentFormat) {
    bool _validDate = false;

    if (date is String) {
      if (_regularExpression(date)) {
        _validDate = true;
        List<String> _dateParts = date.split(RegExp(r'[-\\\/]'));

        _validDate = _validStringMonth(_dateParts[1]) &&
            _validStringDay(_dateParts, currentFormat);
      }
    }

    return _validDate;
  }

  bool validDateTime(date) {
    bool _valid = false;

    if (date is DateTime) {
      _valid = _validStringMonth(date.month) &&
          _validStringDay([date.day, date.month], 'dmy');
    }

    return _valid;
  }

  bool _validInt(date) {
    bool _valid = false;

    if (date is int) {
      _valid = true;
    }

    return _valid;
  }

  bool _validLocale(locale) => DateFormat.localeExists(locale);

  assingExpectedFormat(expectedTypeOrFormat) {
    String _availableFormat = 'yMd';

    if (expectedTypeOrFormat == String) {
      expectedTypeOrFormat = _availableFormat;
    }

    return expectedTypeOrFormat;
  }

  void assingLocale(locale) {
    String _locale = 'en_US';

    initializeDateFormatting(_locale, null);
    Intl.defaultLocale = _locale;

    if (_validLocale(locale)) {
      _locale = locale;

      initializeDateFormatting(_locale, null);
      Intl.defaultLocale = _locale;
    }
  }

  bool _validexpectedTypeOrFormat({expectedTypeOrFormat, currentStringFormat}) {
    bool _valid = false;

    if (expectedTypeOrFormat != null) {
      if (_validDateTypes(expectedTypeOrFormat)) {
        _valid = true;
        if (!(expectedTypeOrFormat == String &&
            validFormats.contains(expectedTypeOrFormat))) {
          _valid = true;
        }
      }
    }

    if (currentStringFormat != null) {
      if (currentStringFormat.runtimeType == String &&
          ['mdy', 'dmy', 'ymd']
              .contains(currentStringFormat.toString().toLowerCase())) {
        _valid = true;
      }
    } else {
      _valid = true;
    }

    return _valid;
  }

  bool _validDateTypes(dateOrDateType) {
    List _generalValidFormats = [DateTime, String, int];

    return _generalValidFormats.contains(dateOrDateType.runtimeType) ||
        dateOrDateType == String;
  }

  bool _validDateByType(date, currentStringFormat) {
    Map<dynamic, bool> _treatmentByTypes = {
      DateTime: validDateTime(date),
      String: _validString(date, currentStringFormat),
      int: _validInt(date),
    };

    return _treatmentByTypes[date.runtimeType]!;
  }

  bool validDate(date, expectedTypeOrFormat, currentStringFormat, locale) {
    bool _valid = false;

    if (_validexpectedTypeOrFormat(
        expectedTypeOrFormat: expectedTypeOrFormat)) {
      _valid = true;
    } else {
      throw ArgumentError('Invalid expected format');
    }

    if (!_validexpectedTypeOrFormat(currentStringFormat: currentStringFormat)) {
      throw ArgumentError('Invalid current string format');
    }

    if (!_validDateTypes(date)) {
      throw ArgumentError(
          'Invalid date type. Allowed formats are: DateTime, String, int');
    }

    if (!_validDateByType(date, currentStringFormat)) {
      throw ArgumentError('The date entered is invalid');
    }

    return _valid;
  }
}


lib/src/parser.dart
import 'package:flutter_date_convert/flutter_date_convert.dart';
import 'package:flutter_date_convert/src/convert.dart';

class Parser {
  Format _format = Format();
  Convert _convert = Convert();

  /// Receives a date as input in the format [DateTime], [int], or [String], and returns the converted date in the desired data type
  ///
  /// Throws an exception if an invalid value is provided for any parameter
  ///
  /// If the input and output types are the same, no conversion will occur, except for [String] to [String] when the input date format is different from the expected format:
  ///
  /// ```dart
  /// Parser _parser = Parser();
  ///
  /// Widget build(BuildContext context) {
  ///   return Text(
  ///     _parser.dateConvert(
  ///       '25/06/2023',
  ///       expectedTypeOrFormat: 'yMMMMd',
  ///       currentStringFormat: 'dmy',
  ///     ),
  ///     textAlign: TextAlign.center,
  ///   );
  /// }
  /// ```
  ///
  /// When converting a date to [String], the [expectedTypeOrFormat] parameter can be either of type [String]
  /// or a String specifying the accepted formats, with the latter option being preferred:
  ///
  /// ```dart
  /// Parser _parser = Parser();
  ///
  /// Widget build(BuildContext context) {
  ///   return Text(
  ///     _parser.dateConvert(
  ///       '25/06/2023',
  ///       expectedTypeOrFormat: String, // 'yMMMMd'
  ///       currentStringFormat: 'dmy',
  ///     ),
  ///     textAlign: TextAlign.center,
  ///   );
  /// }
  /// ```
  ///
  /// The [currentStringFormat] parameter is REQUIRED when the input date is of type [String]
  /// Dates in other formats will not be affected by this parameter
  /// The following options are valid to inform in this parameter: 'dmy', 'mdy', 'ymd'
  /// Providing the incorrect order of day, month, and year in the String may result in an incorrect output or an exception
  ///
  /// The [locale] parameter is optional and can be provided when the output type is [String]. Different output types will not be affected by this parameter
  /// If nothing is provided, the default [locale] value is assumed to be 'en_US'
  /// To identify if a [locale] is valid, refer to: https://pub.dev/documentation/intl/latest/date_symbol_data_http_request/availableLocalesForDateFormatting.html
  ///
  dateConvert(
    date, {
    required expectedTypeOrFormat,
    currentStringFormat,
    String locale = 'en_US',
  }) {
    _format.assingLocale(locale);

    if (_format.validDate(
        date, expectedTypeOrFormat, currentStringFormat, locale)) {
      date = _convert.byExpected(
          date, expectedTypeOrFormat, currentStringFormat, locale);
    }

    return date;
  }
}


lib/src/convert.dart
import 'package:flutter_date_convert/flutter_date_convert.dart';
import 'package:intl/intl.dart';

class Convert {
  Format _format = Format();

  _convertDateTimeToTimestamp(date, expectedTypeOrFormat) {
    if (date is DateTime && expectedTypeOrFormat is int) {
      date = date.millisecondsSinceEpoch;
    }

    return date;
  }

  _convertDateTimeToString(date, expectedTypeOrFormat, locale) {
    if (date is DateTime &&
        _format.typeOrFormatIsString(expectedTypeOrFormat)) {
      expectedTypeOrFormat = _format.assingExpectedFormat(expectedTypeOrFormat);

      date = DateFormat(expectedTypeOrFormat).format(date);
    }

    return date;
  }

  _convertStringToDateTime(date, expectedTypeOrFormat) {
    if (date is String && expectedTypeOrFormat == DateTime) {
      date = DateTime.parse(date.split(RegExp(r'[-\\\/]')).reversed.join());
    }

    return date;
  }

  _convertStringToString(date, expectedTypeOrFormat, locale) {
    if (date is String && _format.typeOrFormatIsString(expectedTypeOrFormat)) {
      expectedTypeOrFormat = _format.assingExpectedFormat(expectedTypeOrFormat);

      date = _convertStringToDateTime(date, DateTime);

      date = _convertDateTimeToString(date, expectedTypeOrFormat, locale);
    }

    return date;
  }

  _convertStringToTimestamp(date, expectedTypeOrFormat) {
    if (date is String && expectedTypeOrFormat is int) {
      date = _convertStringToDateTime(date, expectedTypeOrFormat);

      date = _convertDateTimeToTimestamp(date, expectedTypeOrFormat);
    }

    return date;
  }

  _convertTimestampToDateTime(date, expectedTypeOrFormat) {
    if (date is int && expectedTypeOrFormat == DateTime) {
      date = DateTime.fromMillisecondsSinceEpoch(date);
    }

    return date;
  }

  _convertTimestampToString(date, expectedTypeOrFormat, locale) {
    if (date is int && _format.typeOrFormatIsString(expectedTypeOrFormat)) {
      expectedTypeOrFormat = _format.assingExpectedFormat(expectedTypeOrFormat);
      date = _convertTimestampToDateTime(date, DateTime);

      date = _convertDateTimeToString(date, expectedTypeOrFormat, locale);
    }

    return date;
  }

  _fromDateTime(date, expectedTypeOrFormat, locale) {
    Map<dynamic, dynamic> _treatmentByFormat = {
      int: _convertDateTimeToTimestamp(date, expectedTypeOrFormat),
      String: _convertDateTimeToString(date, expectedTypeOrFormat, locale),
    };

    return _treatmentByFormat[expectedTypeOrFormat];
  }

  _fromString(date, expectedTypeOrFormat, locale) {
    var _typeOrFormat = expectedTypeOrFormat.runtimeType == String
        ? String
        : expectedTypeOrFormat;

    Map<dynamic, dynamic> _treatmentByFormat = {
      DateTime: _convertStringToDateTime(date, expectedTypeOrFormat),
      String: _convertStringToString(date, expectedTypeOrFormat, locale),
      int: _convertStringToTimestamp(date, expectedTypeOrFormat),
    };

    return _treatmentByFormat[_typeOrFormat];
  }

  _fromTimestamp(date, expectedTypeOrFormat, locale) {
    var _typeOrFormat = expectedTypeOrFormat.runtimeType == String
        ? String
        : expectedTypeOrFormat;

    Map<dynamic, dynamic> _treatmentByFormat = {
      DateTime: _convertTimestampToDateTime(date, expectedTypeOrFormat),
      String: _convertTimestampToString(date, expectedTypeOrFormat, locale),
    };

    return _treatmentByFormat[_typeOrFormat];
  }

  fromInt(date) {
    int? _convertedDate;

    if (date != null) {
      if (date is int) {
        _convertedDate = date;
      }

      if (date is String) {
        _convertedDate = int.parse(date);
      }
    }

    return _convertedDate;
  }

  byExpected(date, expectedTypeOrFormat, currentStringFormat, locale) {
    var _convertedDate = date;

    if (date is DateTime && expectedTypeOrFormat != DateTime) {
      _convertedDate = _fromDateTime(date, expectedTypeOrFormat, locale);
    }

    if (date is String) {
      _convertedDate = _fromString(date, expectedTypeOrFormat, locale);
    }

    if (date is int && expectedTypeOrFormat != int) {
      _convertedDate = _fromTimestamp(date, expectedTypeOrFormat, locale);
    }

    return _convertedDate;
  }
}


lib/src/utils.dart
int hoursInDay = 24;
int milissegundosEmUmSegundo = 1000;
int secondsInHour = 3600;

Map<int, dynamic> mounthsNonLeapYear = {
  1: {
    'name': 'January',
    'days': 31,
  },
  2: {
    'name': 'February',
    'days': 28,
  },
  3: {
    'name': 'March',
    'days': 31,
  },
  4: {
    'name': 'April',
    'days': 30,
  },
  5: {
    'name': 'May',
    'days': 31,
  },
  6: {
    'name': 'June',
    'days': 30,
  },
  7: {
    'name': 'July',
    'days': 31,
  },
  8: {
    'name': 'August',
    'days': 31,
  },
  9: {
    'name': 'September',
    'days': 30,
  },
  10: {
    'name': 'October',
    'days': 31,
  },
  11: {
    'name': 'November',
    'days': 30,
  },
  12: {
    'name': 'December',
    'days': 31,
  },
};

List<String> validFormats = [
  'd',
  'E',
  'EEEE',
  'LLL',
  'LLLL',
  'M',
  'Md',
  'MEd',
  'MMM',
  'MMMd',
  'MMMEd',
  'MMMM',
  'MMMMd',
  'MMMMEEEEd',
  'QQQ',
  'QQQQ',
  'y',
  'yM',
  'yMd',
  'yMEd',
  'yMMM',
  'yMMMd',
  'yMMMEd',
  'yMMMM',
  'yMMMMd',
  'yMMMMEEEEd',
  'yQQQ',
  'yQQQQ',
  'H',
  'Hm',
  'Hms',
  'j',
  'jm',
  'jms',
  'm',
  'ms',
  's',
  'yyyy-MM-dd hh:mm',
  'yyyy-MM-dd hh:mm:ss',
  'yyyy-mm-dd hh:mm',
  'yyyy-mm-dd hh:mm:ss',
  'jmv',
  'jmz',
  'jz'
];
